### PR TITLE
Add UI Materials

### LINKED ISSUE #5607: Add support for applying a shader to a UI Image
## What problem does this solve or what need does it fill?

Being able to apply shaders to UI Images could open a lot of possibilities for nice UIs! Because as of right now I feel like we're a bit limited when we want to make animated UI components.

## What solution would you like?

With the current way to add a shader to an Image, being able to insert a `MaterialMesh2dBundle` to a UI could work!

## What alternative(s) have you considered?

Not sure about this one, but if there was a way to link a fragment shader to an Image directly, there would maybe be no need to change anything for the UI Images?

### PR DESCRIPTION
**Pull request by MarkusTheOrt**
# Objective

- Add Ui Materials so that UI can render more complex and animated widgets.
- Fixes #5607 

## Solution
- Create a UiMaterial trait for specifying a Shader Asset and Bind Group Layout/Data.
- Create a pipeline for rendering these Materials inside the Ui layout/tree.
- Create a MaterialNodeBundle for simple spawning.

## Changelog

- Created a `UiMaterial` trait for specifying a Shader asset and Bind Group.
- Created a `UiMaterialPipeline` for rendering said Materials.
- Added Example [`ui_material` ](https://github.com/MarkusTheOrt/bevy/blob/ui_material/examples/ui/ui_material.rs) for example usage.
- Created [`UiVertexOutput`](https://github.com/MarkusTheOrt/bevy/blob/ui_material/crates/bevy_ui/src/render/ui_vertex_output.wgsl) export as VertexData for shaders.
- Created [`material_ui`](https://github.com/MarkusTheOrt/bevy/blob/ui_material/crates/bevy_ui/src/render/ui_material.wgsl) shader as default for both Vertex and Fragment shaders.

### CODE REVIEW

**Comment by IceSentry** [comment]
Could you add a simple example showing how to use this? I can figure it out easily but I believe this feature deserve a standalone example.

**Comment by MarkusTheOrt** [comment]
I just added a simple example showcasing the feature. 
I am thinking about ditching the use of `UiPipeline` inside the `UiMaterialPipeline` so that I can get rid of the default image bindings. Not every shader requires a texture, and if one is required it can be passed using the material.

Next up I do some renaming and documenting

**Comment by MarkusTheOrt** [comment]
I did now remove the usage of `UiPipeline`, and copied the relevant code into the `UiMaterialPipeline`.
I also did split up the additions, with essentially only the `UiMaterial` trait and its accompanying `UiMaterialKey` being inside the `material.rs` and the rest being in `render/material_pipeline.rs`.

File: crates/bevy_ui/src/node_bundles.rs
Comment (ickshonpe) [inline crates/bevy_ui/src/node_bundles.rs:330]: Seems back to front, the node doesn't render the shader.
```suggestion
/// A UI node that is rendered using a [`UiMaterial`]
```

File: crates/bevy_ui/src/node_bundles.rs
Comment (ickshonpe) [inline crates/bevy_ui/src/node_bundles.rs:338]:
```suggestion
    /// The [`UiMaterial`] used to render the node.
```

File: examples/ui/ui_material.rs
Comment (ickshonpe) [inline examples/ui/ui_material.rs:1]:
```suggestion
//! This example illustrates how to create a Node that shows a Material and derives parameters from
```

"driver" meant to be "derives"?

File: crates/bevy_ui/src/render/material_pipeline.rs
Comment (ickshonpe) [inline crates/bevy_ui/src/render/material_pipeline.rs:103]:
```suggestion
pub struct UiMaterialMeta<M: UiMaterial> {
```

Intuitively, `Mat` implies something to do with matrices. It's obviously not once you look at the type but could be confusing at a glance.

File: crates/bevy_ui/src/render/material_pipeline.rs
Comment (ickshonpe) [inline crates/bevy_ui/src/render/material_pipeline.rs:120]: Same with this, I'd go with:
```suggestion
pub struct UiMaterialBatch<M: UiMaterial> {
```

File: crates/bevy_ui/src/render/material_pipeline.rs
Comment (ickshonpe) [inline crates/bevy_ui/src/render/material_pipeline.rs:280]:
```suggestion
pub struct DrawUiMaterialNode<M>(PhantomData<M>);
```

File: crates/bevy_ui/src/render/material_pipeline.rs
Comment (ickshonpe) [inline crates/bevy_ui/src/render/material_pipeline.rs:462]: If `mode` is always going to be set to `0`, shouldn't we add a separate `UiMaterialVertex` with `mode` removed? The batching optimization it allows can't be implemented with this design anyway (and probably aren't necessary).

Comment (MarkusTheOrt) [reply crates/bevy_ui/src/render/material_pipeline.rs]: I did just as you suggested, had to make some other changes as well 👍 

thanks for the suggestion!

File: examples/ui/ui_material.rs
Comment (MarkusTheOrt) [reply examples/ui/ui_material.rs:1]: I wanted to write "drives" as in "one can use a system that drives the parameters of this material".

I don't think its a great description, but can't seem to find the words for a better one right now.

File: crates/bevy_ui/src/node_bundles.rs
Comment (MarkusTheOrt) [reply crates/bevy_ui/src/node_bundles.rs:329]: Great suggestion, I agree with you there, I wasn't aware of `ContentSize`

I'd like to get this PR done and would be happy to take on the `MaterialImageBundle` as a followup once it is ready

**Comment by MarkusTheOrt** [comment]
I did Change some of the names like @ickshonpe suggested.

I am thinking that it would be a good addition to add the [bevy_render::globals](https://github.com/bevyengine/bevy/blob/main/crates/bevy_render/src/globals.rs) into the default shader bindings.
I believe it is a sensible default since `time` can be useful when doing anything animated within a shader.

File: assets/shaders/circle_shader.wgsl
Comment (st0rmbtw) [inline assets/shaders/circle_shader.wgsl:4]: Maybe move the VertexOutput into an import to make less boilerplate?

Comment (MarkusTheOrt) [reply assets/shaders/circle_shader.wgsl:4]: I think this is the right approach. Should I just add the export at the top of the current shader, or should I create a new file that only houses the struct itself that has the export?

Comment (st0rmbtw) [reply assets/shaders/circle_shader.wgsl:4]: @MarkusTheOrt Yep, that would be nice!

Comment (MarkusTheOrt) [reply assets/shaders/circle_shader.wgsl]: I added the output inside the `material.wgsl` shader and have also implemented it in the `circle_shader.wgsl` 👍

Comment (st0rmbtw) [reply assets/shaders/circle_shader.wgsl]: @MarkusTheOrt Oh, sorry, I misread your comment. I think that a new file `ui_vertex_output.wgsl` that contains only `UiVertexOutput` would be better, like it's done in https://github.com/bevyengine/bevy/blob/main/crates/bevy_sprite/src/mesh2d/mesh2d_vertex_output.wgsl

**Comment by ickshonpe** [comment]
> > The issue is that prepare_uinodes and prepare_uimaterial_nodes are working independently of each other.
> 
> I am not sure if there is anything that can be done here. Disabling all batching is sub-optimal to say the least. Maybe there can be batching for each _layer_ in the Ui?

Batching doesn't have to be disabled, that was just a quick hack to diagnose the problem. There are a couple of approaches that could work, maybe a shared list resource that keeps track of what has been extracted at each stack index so the prepare systems know when to break up the batches. Or just one prepare system that handles batching for all the different extracted items.

File: assets/shaders/circle_shader.wgsl
Comment (st0rmbtw) [reply assets/shaders/circle_shader.wgsl]: Thank you! :)

**Comment by MarkusTheOrt** [comment]
> Batching doesn't have to be disabled, that was just a quick hack. There are a couple of approaches that could work, maybe a shared list resource that keeps track of what has been extracted at each stack index so the prepare systems know when to break up the batches. Or just one prepare system that handles batching for all the different extracted items.

I've been thinking about this, maybe a marker component for Material Nodes that can be queried for in `extract_uinodes` and then used in `prepare_uinodes` can be used.

**Comment by ickshonpe** [comment]
> EDIT: I've tried doing that and it seems to be working from a batching standpoint, looking at it in renderdoc it does two draw calls, one for the black square, borders, and the vertical bar, and one for the red bar, the circle still draws on top (I guess that is what your PR (#9062 ) fixes?

Sounds like all you need for it to work is to set the `z` field of each `UiBatch` and `UiMaterialBatch` equal to the stack index (`as f32`) of the last extracted UI node belonging to each batch, except for maybe some local ordering issues for nodes with more than one extracted item.

**Comment by MarkusTheOrt** [comment]
> Yes, sounds like you are almost there. All you should need to do is set the `z` field of each `UiBatch` and `UiMaterialBatch` equal to the stack index (`as f32`) of the last extracted UI node belonging to each batch. Maybe there might be some local ordering issues for nodes with more than one extracted item but that's not a big deal.

Got it working through that, thank you, I guess i'll make it clear in the initial post that this is blocked until the above mentioned PR is being merged 👍 and then do some final review myself before marking it ready ❤️ thanks for all the feedback and help so far!
Screenshot

**Comment by MarkusTheOrt** [comment]
Okay, I did look over the code quite a bit now and I believe this is ready to merge, just waiting for some feedback and reviews now! 👍

File: crates/bevy_ui/src/render/material_pipeline.rs
Comment (MarkusTheOrt) [inline crates/bevy_ui/src/render/material_pipeline.rs:719]: Should be `extracted_uinode`

File: crates/bevy_ui/src/material.rs
Comment (IceSentry) [inline crates/bevy_ui/src/material.rs:5]:
```suggestion

/// Materials are used alongside [`UiMaterialPlugin`](crate::UiMaterialPipeline) and [`MaterialNodeBundle`](crate::prelude::MaterialNodeBundle)
```

**Review by IceSentry — APPROVED** [review]
This seems really good to me. 

One thing though. Could you rename material and material_pipeline to ui_material and ui_material_pipeline (or alternatively ui_material_plugin). It will make it easier to navigate to the file from an editor. material_pipeline is a bit too generic in my opinion. 

It would also be good to document in that the view binding is present probably somewhere in the UitMaterial doc.

I think there should be a follow up PR that adds at the very least the `globals` binding but maybe something like mouse position too. It doesn't need to be in this PR though. It's just that I'm sure people will use this to make a shadertoy in bevy since it would be the easiest way to do that.

Despite the couple of nitpicks I mentioned I think this could be merged and improved upon after so I'll approve anyway.

File: crates/bevy_ui/src/render/material.wgsl
Comment (MarkusTheOrt) [reply crates/bevy_ui/src/render/material.wgsl]: Now that you mention it, I don't think so. I think its best to remove it, and if there is a use case where vertex color makes sense then we can still add it back with a PR, its not a big change.

File: assets/shaders/circle_shader.wgsl
Comment (MarkusTheOrt) [reply assets/shaders/circle_shader.wgsl:12]: I agree, I've updated the example, the shader now only takes in one property (color) and just draws a simple circle with the color shift.

**Comment by ickshonpe** [comment]
Not forgotten about this, I'll try and do a review a bit later today after I've finished what I'm currently working on.

**Review by alice-i-cecile — APPROVED** [review]
Straightforward and appears to be consistent with how our other materials work. The code quality and docs are good (or at least, on par with the rest of rendering) and this is a very useful feature.

**Comment by nicopap** [comment]
@superdump You were complaining that we were duplicating the material code between 2d and 3d, this would add another instance of such duplication.

**Comment by superdump** [comment]
Indeed. I really want to consolidate 2D and 3D materials and minimise the differences else developing any improvements requires implementing the same thing multiple times.

File: Cargo.toml
Comment (mockersf) [inline Cargo.toml:2329]:
```suggestion
description = "Demonstrates creating and using custom Ui materials"
```

File: crates/bevy_ui/src/render/mod.rs
Comment (mockersf) [inline crates/bevy_ui/src/render/mod.rs:627]:
```suggestion
pub(crate) const QUAD_VERTEX_POSITIONS: [Vec3; 4] = [
```
or document it

File: crates/bevy_ui/src/render/mod.rs
Comment (mockersf) [inline crates/bevy_ui/src/render/mod.rs:634]:
```suggestion
pub(crate) const QUAD_INDICES: [usize; 6] = [0, 2, 3, 0, 1, 2];
```
or document it

File: crates/bevy_ui/src/render/ui_material_pipeline.rs
Comment (mockersf) [inline crates/bevy_ui/src/render/ui_material_pipeline.rs:142]:
```suggestion
// in this [`UiMaterialPipeline`] there is (currently) no batching going on.
// Therefore the [`UiMaterialBatch`] is more akin to a draw call.
```
This seems to be more a code comment than a doc

**Comment by mockersf** [comment]
I think the border info should be part of the `UiVertexOutput` as it's the node responsibility to draw them

**Comment by MarkusTheOrt** [comment]
> I think the border info should be part of the `UiVertexOutput` as it's the node responsibility to draw them

I don't quite understand what you mean.
Don't borders get extracted and rendered just as colored ui nodes?

File: crates/bevy_ui/src/render/ui.wgsl
Comment (superdump) [inline crates/bevy_ui/src/render/ui.wgsl:42]: This change should be removed. :)

Comment (MarkusTheOrt) [reply crates/bevy_ui/src/render/ui.wgsl]: I might have made a complete fool out of myself here :P 
but is this a windows line ending or was that just a joke that completely went past myself?

**Comment by alice-i-cecile** [comment]
No longer controversial, but [waiting on review from @superdump](https://discord.com/channels/691052431525675048/1161381998925991986/1165065824030179448).

**Comment by mockersf** [comment]
> > I think the border info should be part of the `UiVertexOutput` as it's the node responsibility to draw them
> 
> I don't quite understand what you mean. Don't borders get extracted and rendered just as colored ui nodes?

Yes, if you add a `BorderColor` component.

With this PR, the material can be drawn over the border, but we don't know that as it's given the size of the node, border included. I would like to see the border width on each side as part of the data passed to the shader so that the material can decide how to render its borders

**Comment by MarkusTheOrt** [comment]
Okay, I have added the border sizes (in UV _size_ if it can be called that) 
so if you have a 100px square with a 10px border the values passed will be [0.1, 0.1, 0.1, 0.1] in the order of [left, right, top. bottom]

File: crates/bevy_ui/src/node_bundles.rs
Comment (superdump) [reply crates/bevy_ui/src/node_bundles.rs:345]: Couldn’t the image just be a member of the material type? @ickshonpe

File: crates/bevy_ui/src/render/ui_material_pipeline.rs
Comment (superdump) [inline crates/bevy_ui/src/render/ui_material_pipeline.rs:51]:
```suggestion
pub const UI_MATERIAL_SHADER_HANDLE: Handle<Shader> = Handle::weak_from_u128(10074188772096983955);
```

File: crates/bevy_ui/src/render/ui_material_pipeline.rs
Comment (superdump) [inline crates/bevy_ui/src/render/ui_material_pipeline.rs:119]: Does this need to be per material or would one for all materials + instance offsets be enough?

File: crates/bevy_ui/src/render/ui_material_pipeline.rs
Comment (superdump) [inline crates/bevy_ui/src/render/ui_material_pipeline.rs:220]:
```suggestion
            label: Some("ui_material_pipeline".into()),
```

File: crates/bevy_ui/src/render/ui_material_pipeline.rs
Comment (MarkusTheOrt) [reply crates/bevy_ui/src/render/ui_material_pipeline.rs:119]: I don’t think so. Should I initialize this one in the normal zu initialization or is there another way to do it once?

**Comment by MarkusTheOrt** [comment]
I’ll resolve these as soon as possible (in about one hour or so)

File: crates/bevy_ui/src/render/ui_material_pipeline.rs
Comment (MarkusTheOrt) [reply crates/bevy_ui/src/render/ui_material_pipeline.rs]: Apologies for the obviously bad question, I did read up on it and I updated everything. Now uses only one UI Meta (like the normal UI does)

File: crates/bevy_ui/src/render/ui_vertex_output.wgsl
Comment (JMS55) [inline crates/bevy_ui/src/render/ui_vertex_output.wgsl:6]:
```suggestion
    // The size of the borders in UV space. Order is Left, Right, Top, Bottom.
```

File: crates/bevy_ui/src/ui_material.rs
Comment (JMS55) [inline crates/bevy_ui/src/ui_material.rs:70]:
```
/// Also note that binding group 0 is always taken with the view matrix.
```

This is a weird sentence, not sure what it's trying to say.

File: crates/bevy_ui/src/render/ui_material_pipeline.rs
Comment (JMS55) [inline crates/bevy_ui/src/render/ui_material_pipeline.rs:25]: Replace with a * import, see https://github.com/bevyengine/bevy/pull/9847

File: crates/bevy_ui/src/render/ui_material_pipeline.rs
Comment (JMS55) [inline crates/bevy_ui/src/render/ui_material_pipeline.rs:56]: The last part of this sentence is confusing to me.

**Review by JMS55 — APPROVED** [review]
Code looks generally fine, although I'm not familiar with any of the UI stuff. Some small doc nits.

File: crates/bevy_ui/src/ui_material.rs
Comment (MarkusTheOrt) [reply crates/bevy_ui/src/ui_material.rs:70]: this is supposed to say that bind group 0 (in wgsl `@group(0)`) is always bound to the View matrix from `bevy_render`.

from the ui shader:
```wgsl
#import bevy_render::view::View

@group(0) @binding(0)
var<uniform> view: View;
```

File: crates/bevy_ui/src/render/ui_material_pipeline.rs
Comment (MarkusTheOrt) [reply crates/bevy_ui/src/render/ui_material_pipeline.rs:38]: I removed this part as it is confusing and doesn't convey any extra information. (I copied this from the Mesh2d Material pipeline. 👍

File: crates/bevy_ui/src/ui_material.rs
Comment (MarkusTheOrt) [reply crates/bevy_ui/src/ui_material.rs]: I changed the wording a bit and added a link to the View Uniform: 
`/// Also note that bind group 0 is always bound to the [`View Uniform`](bevy_render::view::ViewUniform).`

**Comment by alice-i-cecile** [comment]
I'm less sleep-deprived now and feeling comfortable re-reviewing this. I'm happy with the quality of code, the docs are great, and this is useful abstraction. Merging for 0.12 :)

**Pull request merged**
